posthog/web_vitals/sql.py

```diff
@@ -0,0 +1,125 @@
+from django.conf import settings
+
+from posthog.clickhouse.kafka_engine import kafka_engine, ttl_period
+from posthog.clickhouse.table_engines import (
+    Distributed,
+    ReplicationScheme,
+    MergeTreeEngine,
+)
+from posthog.kafka_client.topics import KAFKA_CLICKHOUSE_WEB_VITALS_EVENTS
+
+WEB_VITALS_DATA_TABLE = lambda: "sharded_web_vitals"
+
+
+KAFKA_WEB_VITALS_TABLE_BASE_SQL = """
+CREATE TABLE IF NOT EXISTS {table_name} ON CLUSTER '{cluster}'
+(
+    session_id VARCHAR,
+    team_id Int64,
+    timestamp DateTime64(6, 'UTC'),
+    current_url VARCHAR,
+    fcp Nullable(Float64),
+    lcp Nullable(Float64),
+    cls Nullable(Float64),
+    inp Nullable(Float64),
+    -- we store the rest of the properties as a JSON string, just like the events table
+    properties VARCHAR CODEC(ZSTD(3)),
+) ENGINE = {engine}
+"""
+
+WEB_VITALS_TABLE_BASE_SQL = """
+CREATE TABLE IF NOT EXISTS {table_name} ON CLUSTER '{cluster}'
+(
+    session_id VARCHAR,
+    team_id Int64,
+    timestamp DateTime64(6, 'UTC'),
+    current_url VARCHAR,
+    fcp Nullable(Float64),
+    lcp Nullable(Float64),
+    cls Nullable(Float64),
+    inp Nullable(Float64),
+    properties VARCHAR CODEC(ZSTD(3)),
+    _timestamp DateTime,
+    _offset UInt64,
+    _partition UInt64
+) ENGINE = {engine}
+"""
+
+WEB_VITALS_DATA_TABLE_ENGINE = lambda: MergeTreeEngine("web_vitals", replication_scheme=ReplicationScheme.SHARDED)
+
+WEB_VITALS_TABLE_SQL = lambda: (
+    WEB_VITALS_TABLE_BASE_SQL
+    + """
+    PARTITION BY toYYYYMM(timestamp)
+    ORDER BY (team_id,  toDate(timestamp), current_url)
+    {ttl_period}
+"""
+).format(
+    table_name=WEB_VITALS_DATA_TABLE(),
+    cluster=settings.CLICKHOUSE_CLUSTER,
+    engine=WEB_VITALS_DATA_TABLE_ENGINE(),
+    ttl_period=ttl_period("timestamp", 366, unit="DAY"),
```

Comment:
this is IMHO a relatively long ttl but we could have a roll-up of older data if folk really want longer